### PR TITLE
Large refactor, addition of Task Manager and conversion to PSQ

### PR DESCRIPTION
**Pull request by aarontp**
This refactors a large amount of the Turbinia code and touches almost everything.  It's still a WiP, but the major refactoring is done and the general flow works end to end.  Here are some of the major changes:
- Converts Turbinia from Celery to using PSQ for managing tasks
- Implements a Task Manager to manage new evidence to process and creates and manages PSQ tasks.
- Adds usage of different Evidence types
- Feeds newly created output evidence from Tasks back into Turbinia to be re-processed.
- Turbiniactl can add different evidence types, start workers, list jobs and by default (for now) will run the Task Manager indefinitely.

Caveats:
- Plaso/BE Tasks are broken 
- Still lots of TODOs spread throughout the code
- Some PubSub code is there, but a lot of it references a previous iteration that was planning to use native PubSub for all task management.  Should probably be mostly ignored for the purposes of this PR, and I will clean it up later.

### CODE REVIEW

File: turbinia/__init__.py [hunk 1]
```diff
@@ -17,34 +17,19 @@
 import os
 import sys
 
-from celery import Celery
-
 from turbinia import config
 
-VERSION = '20150916'
+log = logging.getLogger('turbinia')
```
Comment (berggren) [inline turbinia/__init__.py]: Just to be constant, use u'' instead.

File: turbinia/__init__.py [hunk 2]
```diff
@@ -17,34 +17,19 @@
...
+
+class TurbiniaException(Exception):
+  pass
 
 try:
   config.LoadConfig()
 except config.TurbiniaConfigException as e:
   # pylint: disable=logging-format-interpolation
-  logging.fatal('Could not load Turbinia config: {0:s}'.format(str(e)))
+  log.fatal('Could not load Turbinia config: {0:s}'.format(str(e)))
```
Comment (berggren) [inline turbinia/__init__.py]: Same here, use u'' througout the codebase. Easier to catch unicode problems own the read.

Comment (berggren) [inline turbinia/__init__.py]: Oh, I see this is the case all over. Let's do this in a followup CL

File: turbinia/config/__init__.py [hunk 3]
```diff
@@ -44,13 +52,18 @@
...
 
+CONFIG = None
 
 class TurbiniaConfigException(Exception):
   pass
 
 
 def LoadConfig():
   """Finds Turbinia config file and loads it."""
+  global CONFIG
```
Comment (berggren) [inline turbinia/config/__init__.py]: Is there a way to handle this without a global?

Comment (aarontp) [inline turbinia/config/__init__.py]: As discussed, possibly with a singleton pattern on the config class.  I put in a TODO for now.

File: turbinia/__init__.py (hunk 2 above)
Comment (aarontp) [inline turbinia/__init__.py]: I put a TODO for this in the TODO.md for now.

File: turbinia/task_manager.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    """Process any tasks that need to be processed.
+
+    Returns:
+      The number of tasks that have completed.
+    """
+    raise NotImplementedError
+
+  def run(self):
+    """Main run loop for TaskManager."""
+    log.info(u'Starting PSQ Task Manager run loop')
```
Comment (berggren) [inline turbinia/task_manager.py]: This is the base class, so I guess drop the PSQ reference here?

File: turbinia/workers/__init__.py [hunk 5]
```diff
@@ -13,53 +13,165 @@
...
+        error: Short string describing the error.
+        traceback: Traceback of the error.
+    """
+    self.error['error'] = error
+    self.error['traceback'] = traceback
 
 
-class TurbiniaTask(Task):
+class TurbiniaTask(object):
   """Base class for Turbinia tasks."""
```
Comment (berggren) [inline turbinia/workers/__init__.py]: Add attributes

File: turbinia/workers/__init__.py [hunk 6]
```diff
@@ -13,53 +13,165 @@
...
-  abstract = True
 
-  def __init__(self):
-    super(TurbiniaTask, self).__init__()
+  def __init__(self, name=None, base_output_dir=None):
```
Comment (berggren) [inline turbinia/workers/__init__.py]: docstring

File: turbinia/workers/__init__.py [hunk 7]
```diff
@@ -13,53 +13,165 @@
...
+    self.id = uuid.uuid4().hex
+    self.name = name if name else self.__class__.__name__
+    self.base_output_dir = base_output_dir
+    self.output_dir = None
+    self.result = None
+
+  def setup(self, evidence):
+    """Perform common setup operations and runtime environment.
```
Comment (berggren) [inline turbinia/workers/__init__.py]: Args

File: turbinia/workers/be.py [hunk 8]
```diff
@@ -49,8 +50,7 @@ def run(src_path, out_path, offsets, job_id, **kwargs):
 class BulkExtractorCalcOffsetsTask(TurbiniaTask):
   """Task to calculate offsets for Bulk extractor."""
 
-  @staticmethod
-  def run(src_path, num_workers, page_size=16777216):
+  def run(self, evidence, num_workers, page_size=16777216):
```
Comment (berggren) [inline turbinia/workers/be.py]: Please  put this magic number in a constant

File: turbinia/workers/be.py [hunk 9]
```diff
@@ -102,8 +102,7 @@ def run(src_path, num_workers, page_size=16777216):
 class BulkExtractorReducerTask(TurbiniaTask):
   """Reduce bulk extractor outputs."""
 
-  @staticmethod
-  def run(results):
+  def run(self, evidence, results):
     """Celery task that reduces the results into one SQLite database.
```
Comment (berggren) [inline turbinia/workers/be.py]: This is not a Celery task anymore, right?

File: turbinia/config/logger.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  # TODO(aarontp): Add a config option to set the log level
+  config.LoadConfig()
+  log = logging.getLogger('turbinia')
+
+  fh = logging.FileHandler(config.LOG_FILE)
+  formatter = logging.Formatter(u'%(asctime)s:%(levelname)s:%(message)s')
+  fh.setFormatter(formatter)
+  fh.setLevel(logging.DEBUG)
+
+  ch = logging.StreamHandler()
```
Comment (aarontp) [inline turbinia/config/logger.py]: Done.

File: turbinia/task_manager.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+def task_runner(obj, *args, **kwargs):
+  """Wrapper function to run specified TurbiniaTask object.
+
+  Args:
+    obj: An instantiated TurbiniaTask object.
+    *args: Any Args to pass to obj.
+    **kwargs: Any keyword args to pass to obj.
+
+  Returns:
+    Output from TurbiniaTask (should be TurbiniaTaskReslt).
```
Comment (aarontp) [inline turbinia/task_manager.py]: done.

File: turbinia/task_manager.py [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    # propogate errors to the Task Manager
+    logging.warning(u'Exception thrown from Task: {0:s}'.format(
+        traceback.format_exc()))
+    raise e
+
+  return res
+
+
+class TaskManager(object):
+  """Class to manage Turbinia Tasks.
```
Comment (aarontp) [inline turbinia/task_manager.py]: Makes sense.  Renamed.

File: turbinia/task_manager.py (hunk 4 above)
Comment (aarontp) [inline turbinia/task_manager.py]: Good point, thanks.

File: turbinia/__init__.py (hunk 1 above)
Comment (aarontp) [inline turbinia/__init__.py]: As mentioned below, TODO in TODO.md.

File: turbinia/workers/__init__.py (hunk 5 above)
Comment (aarontp) [inline turbinia/workers/__init__.py]: Done.

File: turbinia/workers/__init__.py (hunk 6 above)
Comment (aarontp) [inline turbinia/workers/__init__.py]: Done.

File: turbinia/workers/__init__.py (hunk 7 above)
Comment (aarontp) [inline turbinia/workers/__init__.py]: Done.

File: turbinia/workers/be.py (hunk 8 above)
Comment (aarontp) [inline turbinia/workers/be.py]: Done.

File: turbinia/workers/be.py (hunk 9 above)
Comment (aarontp) [inline turbinia/workers/be.py]: Yep.  Removed.  There's still lots of other broken things to fix in this Task, but I'll tackle them later.

File: turbinia/workers/be.py (hunk 8 above)
Comment (aarontp) [inline turbinia/workers/be.py]: (I'm assuming that is what you meant.  Let me know if you meant something else).

**Comment by aarontp** [comment]
Thanks!  PTAL.

**Comment by berggren** [comment]
LGTM

**Pull request merged**